File: web/src/live/ringBuffer.ts
/**
 * Fixed-capacity ring buffer of (deviceTimeUs, value) points for live waveform
 * rendering. Mutated at full sample rate by the packet handlers and read by the
 * canvas on requestAnimationFrame — never via React state, so the UI is not
 * re-rendered per sample (§2.4).
 */
export class WaveBuffer {
  private ts: Float64Array;
  private vs: Float32Array;
  private head = 0;
  private filled = 0;

  constructor(public readonly capacity: number) {
    this.ts = new Float64Array(capacity);
    this.vs = new Float32Array(capacity);
  }

  push(t: number, v: number): void {
    this.ts[this.head] = t;
    this.vs[this.head] = v;
    this.head = (this.head + 1) % this.capacity;
    if (this.filled < this.capacity) this.filled++;
  }

  get length(): number {
    return this.filled;
  }

  latestT(): number | null {
    if (this.filled === 0) return null;
    return this.ts[(this.head - 1 + this.capacity) % this.capacity];
  }

  /** Points with timestamp ≥ tFrom, in chronological order. */
  window(tFrom: number): { t: number[]; v: number[] } {
    const t: number[] = [];
    const v: number[] = [];
    for (let i = 0; i < this.filled; i++) {
      const idx = (this.head - this.filled + i + this.capacity * 2) % this.capacity;
      if (this.ts[idx] >= tFrom) {
        t.push(this.ts[idx]);
        v.push(this.vs[idx]);
      }
    }
    return { t, v };
  }

  clear(): void {
    this.head = 0;
    this.filled = 0;
  }
}

File: ios/VitalView/Views/RingBuffer.swift
//
//  RingBuffer.swift
//  VitalView
//
//  Fixed-capacity ring buffer of (deviceTimeUs, value) points for live waveform
//  rendering. Mutated at full sample rate by the packet handlers and read by the Canvas
//  on a TimelineView tick — never via @Published state, so the UI is not re-rendered per
//  sample.
//
//  Ported from web/src/live/ringBuffer.ts (WaveBuffer). Reads/writes are guarded by a
//  lock since the source delivers packets on a background queue while the Canvas reads
//  on the main thread.
//

import Foundation

final class WaveBuffer {
    let capacity: Int
    private var ts: [Double]
    private var vs: [Double]
    private var head = 0
    private var filled = 0
    private let lock = NSLock()

    init(capacity: Int) {
        self.capacity = capacity
        self.ts = Array(repeating: 0, count: capacity)
        self.vs = Array(repeating: 0, count: capacity)
    }

    func push(_ t: Double, _ v: Double) {
        lock.lock()
        ts[head] = t
        vs[head] = v
        head = (head + 1) % capacity
        if filled < capacity { filled += 1 }
        lock.unlock()
    }

    var length: Int {
        lock.lock(); defer { lock.unlock() }
        return filled
    }

    func latestT() -> Double? {
        lock.lock(); defer { lock.unlock() }
        if filled == 0 { return nil }
        return ts[(head - 1 + capacity) % capacity]
    }

    /// Points with timestamp ≥ tFrom, in chronological order.
    func window(from tFrom: Double) -> (t: [Double], v: [Double]) {
        lock.lock(); defer { lock.unlock() }
        var t: [Double] = []
        var v: [Double] = []
        t.reserveCapacity(filled)
        v.reserveCapacity(filled)
        for i in 0..<filled {
            let idx = (head - filled + i + capacity * 2) % capacity
            if ts[idx] >= tFrom {
                t.append(ts[idx])
                v.append(vs[idx])
            }
        }
        return (t, v)
    }

    func clear() {
        lock.lock()
        head = 0
        filled = 0
        lock.unlock()
    }
}

/// The non-reactive set of waveform ring buffers, mirroring web `live/runtime.ts`.
final class WaveBuffers {
    let ecg = WaveBuffer(capacity: 4096)
    let ppgGreen = WaveBuffer(capacity: 8192)
    let ppgRed = WaveBuffer(capacity: 8192)
    let ppgIr = WaveBuffer(capacity: 8192)
    let biozDz = WaveBuffer(capacity: 2048)
    let z0 = WaveBuffer(capacity: 1024)

    func clear() {
        ecg.clear(); ppgGreen.clear(); ppgRed.clear(); ppgIr.clear(); biozDz.clear(); z0.clear()
    }

    /// Latest device timestamp across the high-rate streams.
    func latestDeviceUs() -> Double {
        max(ecg.latestT() ?? 0, ppgGreen.latestT() ?? 0, biozDz.latestT() ?? 0)
    }
}
